ghost/core/core/server/services/members/jobs/clean-expired-comped.js

```diff
@@ -0,0 +1,72 @@
+const {parentPort} = require('worker_threads');
+const debug = require('@tryghost/debug')('jobs:clean-expired-comped');
+const moment = require('moment');
+
+// recurring job to clean expired complimentary subscriptions
+
+// Exit early when cancelled to prevent stalling shutdown. No cleanup needed when cancelling as everything is idempotent and will pick up
+// where it left off on next run
+function cancel() {
+    if (parentPort) {
+        parentPort.postMessage('Expired complimentary subscriptions cleanup cancelled before completion');
+        parentPort.postMessage('cancelled');
+    } else {
+        setTimeout(() => {
+            process.exit(0);
+        }, 1000);
+    }
+}
+
+if (parentPort) {
+    parentPort.once('message', (message) => {
+        if (message === 'cancel') {
+            return cancel();
+        }
+    });
+}
+
+(async () => {
+    const cleanupStartDate = new Date();
+    const db = require('../../../data/db');
+    debug(`Starting cleanup of expired comp subscriptions`);
+    const expiredCompedRows = await db.knex('members_products')
+        .where('expiry_at', '<', moment.utc().startOf('day').toISOString())
+        .select('*');
+
+    let deletedExpiredSubs = 0;
+    let updatedMembers = 0;
+
+    // Run cleanup for expired comp subscriptions
+    // Removes expired entries and updates affected member statuses
+    if (expiredCompedRows?.length) {
+        const rowIds = expiredCompedRows.map(d => d.id);
+        const memberIds = expiredCompedRows.map(d => d.member_id);
+
+        // Delete all expired comped rows
+        deletedExpiredSubs = await db.knex('members_products')
+            .whereIn('id', rowIds)
+            .del();
+
+        // Update all comped members to free
+        updatedMembers = await db.knex('members')
+            .whereIn('id', memberIds)
+            .andWhere('status', 'comped')
+            .update({
+                status: 'free'
+            });
+    }
+
+    let cleanupEndDate = new Date();
+
+    debug(`Removed ${deletedExpiredSubs} expired subscriptions, updated ${updatedMembers} members in ${cleanupEndDate.valueOf() - cleanupStartDate.valueOf()}ms`);
+
+    if (parentPort) {
+        parentPort.postMessage(`Removed ${deletedExpiredSubs} expired subscriptions, updated ${updatedMembers} members in ${cleanupEndDate.valueOf() - cleanupStartDate.valueOf()}ms`);
```

Comment:
I'm thinking if we would end up with logs full of `Removed 0 expired subscriptions...` on majority of instances :thinking: It is probably fine in the beginning. How do you think we should handle it if that's the case?

Reply:
🤔 Yea that's a fair point, we could remove the logs for when nothing happens in future if that becomes annoying maybe and only output stuff when any changes are done ? Even with optimisations, don't think we can completely avoid not running jobs when there are no expiring subs, so there will always be some case where it runs but doesn't do anything